### PR TITLE
layouts: simple confirmations and warnings

### PR DESCRIPTION
**Pull request by mmilata**
This generalizes the `confirm_action` layout a bit and most of the relevant application code is modified to use it. Same with `show_success`/`show_warning`/`show_error`.

[UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1033168650/artifacts/master_diff/index.html)

### CODE REVIEW

**Comment by andrewkozlik** [comment]
Looking at the UI changes, I can see that the rendering mechanism adds a space after each parameter, e.g. `after 10 seconds?` is now `after 10 seconds ?`, which is incorrect. Hmm, I am guessing that this is due to the fact that `format_parametrized()` strips the spaces and `render_text()` puts them back, which is wrong if no space was stripped. So why do we strip them in the first place?

**Comment by matejcik** [comment]
> why do we strip them in the first place?

because otherwise there will be two spaces.
we can modify `render_text` to not add a space, but that entails reviewing all places that rely on this behavior, which is a lot IIRC

**Comment by andrewkozlik** [comment]
> > why do we strip them in the first place?
> 
> because otherwise there will be two spaces.
> we can modify `render_text` to not add a space, but that entails reviewing all places that rely on this behavior, which is a lot IIRC

So let me rephrase the question. Why do we add the spaces? Do we have any examples that rely on this behavior?

**Comment by mmilata** [comment]
I don't know the original reasoning but at this point I'd say we add the spaces in `render_text` to stay compatible with the previous implementation. Given `text = Text("title"); text.normal("foo", "bar")`, there is going to be either linebreak or space between the two words. As a result there's no way to render two characters of different font next to each other, which would be the right thing to do here.

We probably could change the behavior and update all invocations to be like `text.normal("foo ", "bar")`, however I'd propose to instead solve this by making the `?` part of the parameter, which:
* makes the screen identical to what we currently have in master, and
* does not make the ongoing Rust rewrite of the `render_text` harder.

**Comment by andrewkozlik** [comment]
> We probably could change the behavior and update all invocations to be like `text.normal("foo ", "bar")`, however I'd propose to instead solve this by making the `?` part of the parameter (and thus bold), which:
> 
> * makes the screen identical to what we currently have in master, and
> 
> * does not make the ongoing Rust rewrite of the `render_text` harder.

It seems to me that the correct solution would be not to strip spaces and not to add them at all. That way, if `new_lines=True` then `text.normal("foo", "bar")` remains as is and if `new_lines=False`, then it changes to `text.normal("foo bar")`. I don't think we need to worry about making the screens identical to what we currently have in master, because it's not always correct. Like the question mark, for example, really shouldn't be bold. I can't judge the impact on the ongoing Rust rewrite, but not sure if that should be a factor.

**Comment by mmilata** [comment]
So I removed the spaces in 2febef0 and not much was needed to make UI tests pass: [UI diff from master](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1037645711/artifacts/master_diff/index.html).

Still this change probably introduces some corner case that will break something not covered by test. The code converted to layouts mostly uses `"single strings\nwith embedded\nnewlines"` which is fine, I'll need to review the rest while converting it to layouts.

boldquestionmark

**Comment by mmilata** [comment]
Added two more commits to finish `apps.bitcoin` conversion. [UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1052854587/artifacts/master_diff/index.html)

**Review by tsusanka — APPROVED** [review]
LGTM with the exception of the "require" usage. As discussed on standup it seems safer to have "require" by default and explicitly omit "requiredness" if necessary.

**Comment by matejcik** [comment]
I'd prefer to look this over before it is merged. If I'm blocking something, I'll move this up my priority list :) just let me know

**Comment by mmilata** [comment]
Mind if I rebase this to master?

**Comment by matejcik** [comment]
> Mind if I rebase this to master?

go ahead

**Comment by mmilata** [comment]
Rebased. [UI diff](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1090171496/artifacts/master_diff/index.html)

Compared to master the total allocation count in device tests increased by 42488 (0.3 %). [master](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1089743515/artifacts/core/prof/memperf-html/index.html) -> [this branch](https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/1090171494/artifacts/core/prof/memperf-html/index.html)

File: core/src/trezor/ui/layouts/common.py [hunk 1]
```diff
@@ -29,3 +21,11 @@ async def interact(
     workflow.close_others()
     await ctx.call(ButtonRequest(code=brcode), ButtonAck)
     return await ctx.wait(layout)
+
+
+async def not_cancelled(a: Awaitable[None]) -> bool:
```
Comment (matejcik) [inline core/src/trezor/ui/layouts/common.py:26]: ISTM all uses of `not_cancelled` would be superseded by passing a custom exception instance to `show_error` or `confirm_action`, correct?

File: core/src/trezor/ui/layouts/tt.py [hunk 2]
```diff
@@ -50,58 +60,105 @@
...
+    description_param_font: int = ui.BOLD,
+    verb: Union[str, bytes, None] = Confirm.DEFAULT_CONFIRM,
+    verb_cancel: Union[str, bytes, None] = Confirm.DEFAULT_CANCEL,
+    hold: bool = False,
+    icon: str = None,  # TODO cleanup @ redesign
+    icon_color: int = None,  # TODO cleanup @ redesign
+    reverse: bool = False,  # TODO cleanup @ redesign
+    larger_vspace: bool = False,  # TODO cleanup @ redesign
     br_code: EnumTypeButtonRequestType = ButtonRequestType.Other,
     **kwargs: Any,
```
Comment (matejcik) [inline core/src/trezor/ui/layouts/tt.py:84]: this is now unused

File: core/src/trezor/ui/layouts/tt.py [hunk 3]
```diff
@@ -407,4 +567,67 @@ def confirm_modify_fee(
...
+
+    text = Text("Sign %s" % proto)
+    text.normal(*lines)
+    await raise_if_cancelled(
+        interact(ctx, Confirm(text), "sign_identity", ButtonRequestType.Other)
+    )
+
+
+async def confirm_signverify(
+    ctx: wire.GenericContext, coin: str, message: str, address: str = None
```
Comment (matejcik) [inline core/src/trezor/ui/layouts/tt.py:608]: `address: str = None` seems like an invalid type declaration since some changes in mypy. Is this covered by typecheck?

**Review by matejcik — CHANGES_REQUESTED** [review]
Many comments that are actually one comment: ISTM there is a better alternative to `not_cancelled` function. 

Larger comment: `confirm_action` has way too many parameters now that should go away in the future. I am assuming all the "TODO cleanup" means that you're leaving them in for now for the sake of a clean UI diff?

For the most part though, this is shaping up wonderfully.

File: core/src/trezor/ui/layouts/tt.py (hunk 3 above)
Comment (mmilata) [reply core/src/trezor/ui/layouts/tt.py:608]: It's illegal wrt [PEP484](https://www.python.org/dev/peps/pep-0484/#union-types) but mypy ([still](https://github.com/python/mypy/issues/9091)) allows it. This behavior can be [turned off](https://github.com/trezor/trezor-firmware/blob/ba326865f1b9cacf25eb7d758ba45385cdc41fb7/setup.cfg#L48) but we rely on in in ~300 places. I'll fix the usage in this PR and will look into the rest of the codebase afterwards.

File: core/src/trezor/ui/layouts/tt.py [hunk 4]
```diff
@@ -299,102 +429,132 @@ def show_success(
...
+    description: str = None,
     br_code: EnumTypeButtonRequestType = ButtonRequestType.Other,
-) -> LayoutType:
-    text = Text(title, ui.ICON_SEND, ui.GREEN)
-    text.mono(*_hex_lines(data))
-    return interact(ctx, Confirm(text), br_type, br_code)
+    icon: str = ui.ICON_SEND,  # TODO cleanup @ redesign
+    icon_color: int = ui.GREEN,  # TODO cleanup @ redesign
+    width: int = MONO_HEX_PER_LINE,
+    truncate: bool = False,
```
[the remark this answers is not in the document]
Comment (mmilata) [reply core/src/trezor/ui/layouts/tt.py:471]: It seems to me that there usually isn't a reason to truncate the data, and if there is it probably should be done explicitly. Maybe in the future we should go as far as to make the pagination mandatory whenever a text doesn't fit single screen?

[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/trezor/ui/layouts/tt.py:471]: let's remove the parameter now, and switch to implicitly paginating when that part of code is ready

File: core/src/trezor/ui/layouts/tt.py (hunk 3 above)
Comment (matejcik) [reply core/src/trezor/ui/layouts/tt.py:608]: also worth considering is switching to the new-(new-)style declarations with `str | None`

Comment (matejcik) [reply core/src/trezor/ui/layouts/tt.py:608]: you can configure mypy to assume python 3.10

File: core/src/apps/management/wipe_device.py [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
 from trezor.messages.Success import Success
-from trezor.ui.layouts import confirm_wipe, require
+from trezor.ui.layouts import confirm_wipe
 
 from .apply_settings import reload_settings_from_storage
 
 
 async def wipe_device(ctx, msg):
-    await require(confirm_wipe(ctx))
+    await confirm_wipe(ctx)
```
[the remark this answers is not in the document]
Comment (mmilata) [reply core/src/apps/management/wipe_device.py]: There's some styling missing but since `confirm_action` has everything (w/ kitchen sink) anyway I fixed this in 28f676ca68135ca4a6b46cdf7b18815a361e4b36.

File: core/src/trezor/ui/layouts/common.py (hunk 1 above)
Comment (mmilata) [reply core/src/trezor/ui/layouts/common.py]: Yeah, good idea:+1:. Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

File: core/src/apps/webauthn/add_resident_credential.py [hunk 6]
```diff
@@ -41,14 +41,16 @@ async def add_resident_credential(
...
         cred = Fido2Credential.from_cred_id(bytes(msg.credential_id), None)
     except Exception:
-        text = Text("Import credential", ui.ICON_WRONG, ui.RED)
-        text.normal(
-            "The credential you are",
-            "trying to import does",
-            "not belong to this",
-            "authenticator.",
+        await not_cancelled(
+            show_error(
```
[the remark this answers is not in the document]
Comment (mmilata) [reply core/src/apps/webauthn/add_resident_credential.py]: Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

File: core/src/apps/management/recovery_device/layout.py [hunk 7]
```diff
@@ -23,13 +23,29 @@
 
 async def confirm_abort(ctx: wire.GenericContext, dry_run: bool = False) -> bool:
```
[the remark this answers is not in the document]
Comment (mmilata) [reply core/src/apps/management/recovery_device/layout.py]: Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

File: core/src/apps/common/sdcard.py [hunk 8]
```diff
@@ -1,9 +1,7 @@
 import storage.sd_salt
 from storage.sd_salt import SD_CARD_HOT_SWAPPABLE
 from trezor import fatfs, sdcard, ui, wire
-from trezor.ui.components.tt.text import Text
-
-from apps.common.confirm import confirm, hold_to_confirm
+from trezor.ui.layouts import confirm_action, not_cancelled, show_error
```
[the remark this answers is not in the document]
Comment (mmilata) [reply core/src/apps/common/sdcard.py]: Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

File: core/src/trezor/ui/layouts/tt.py [hunk 9]
```diff
@@ -250,46 +323,103 @@ async def show_address(
...
+        interact(
+            ctx,
+            Confirm(text, confirm=button_confirm, cancel=button_cancel),
+            br_type,
+            br_code,
+        )
+    )
+
+
+async def show_error(
```
[the remark this answers is not in the document]
Comment (mmilata) [reply core/src/trezor/ui/layouts/tt.py]: Went with `show_error_and_raise` but don't feel strongly about it. Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

File: core/src/apps/common/request_pin.py [hunk 10]
```diff
@@ -124,18 +124,30 @@ async def verify_user_pin(
...
 
 async def error_pin_invalid(ctx: wire.Context) -> NoReturn:
-    from apps.common.confirm import confirm
-
-    text = Text("Wrong PIN", ui.ICON_WRONG, ui.RED)
-    text.normal("The PIN you entered is", "invalid.")
-    await confirm(ctx, text, confirm=None, cancel="Close")
+    from trezor.ui.layouts import show_error, not_cancelled
+
+    await not_cancelled(
```
[the remark this answers is not in the document]
Comment (mmilata) [reply core/src/apps/common/request_pin.py]: Fixed in 3c39b4ee1b842dbf9baf26e2491f4697009a1b45.

File: core/src/trezor/ui/layouts/tt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mmilata) [reply core/src/trezor/ui/layouts/tt.py:463]: I'm using the parameter quite a bit in the commits that will follow this PR in order to keep the UI diff minimal. Looking into this.

File: core/src/trezor/ui/layouts/tt.py (hunk 3 above)
Comment (mmilata) [reply core/src/trezor/ui/layouts/tt.py]: Fixed in 4b27e3b0d265ae8a0d16afc78b3c3e688e1e9054.

File: core/src/trezor/ui/layouts/tt.py (hunk 2 above)
Comment (mmilata) [reply core/src/trezor/ui/layouts/tt.py]: Fixed in 28f676ca68135ca4a6b46cdf7b18815a361e4b36.

**Comment by mmilata** [comment]
Thanks for the review @matejcik! Yes, the parameters with the comment are expected to disappear but exist for the sake of changing the UI as little as possible.

Any idea how to verify that the PR doesn't make our fragmentation issues worse? Checking whether individual tests pass/fail on an actual hardware is very time consuming. Perhaps we can run the tests on an emulator repeatedly, decreasing the heap size each iteration until some number of tests start to fail, then compare with master?

File: core/src/trezor/ui/layouts/tt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mmilata) [reply core/src/trezor/ui/layouts/tt.py]: Removed the parameter in 49006de124a5246d58018746cc9bb4117cbb9770, let's discuss it in the next PR with more context.

I think I addressed all your comments, PTAL.

**Comment by matejcik** [comment]
> Perhaps we can run the tests on an emulator repeatedly, decreasing the heap size each iteration until some number of tests start to fail, then compare with master?

this is reasonable. for starters, i'd like to get `test_15_of_15` to fail on emulator, which might give us a reasonable bound for the heap size

also i wouldn't merge this into the upcoming release, which means we will most likely get the workflow-restart thing so that fragmentation is much less of an issue

File: core/src/trezor/ui/layouts/tt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/trezor/ui/layouts/tt.py]: are you ever setting it to `False` explicitly?

File: core/src/apps/common/request_pin.py [hunk 11]
```diff
@@ -124,18 +124,28 @@ async def verify_user_pin(
...
+
+    await show_error_and_raise(
+        ctx,
+        "warning_wrong_pin",
+        header="Wrong PIN",
+        content="The PIN you entered is invalid.",
+        red=True,
+        exc=wire.PinInvalid,
+    )
     raise wire.PinInvalid
```
Comment (matejcik) [inline core/src/apps/common/request_pin.py:137]: isn't this dead code now?

File: core/src/apps/common/request_pin.py [hunk 12]
```diff
@@ -124,18 +124,28 @@ async def verify_user_pin(
...
+
+    await show_error_and_raise(
+        ctx,
+        "warning_invalid_new_pin",
+        header="Invalid PIN",
+        content="The new PIN must be different from your\nwipe code.",
+        red=True,
+        exc=wire.PinInvalid,
+    )
     raise wire.PinInvalid
```
Comment (matejcik) [inline core/src/apps/common/request_pin.py:151]: isn't this dead code now?

File: core/src/trezor/ui/layouts/tt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mmilata) [reply core/src/trezor/ui/layouts/tt.py]: nope

File: core/src/apps/common/sdcard.py [hunk 13]
```diff
@@ -13,61 +11,105 @@ class SdCardUnavailable(wire.ProcessError):
...
+
+async def confirm_retry_sd(
+    ctx: wire.GenericContext,
+    exc: wire.ProcessError = SdCardUnavailable("Error accessing SD card."),
+) -> None:
+    await confirm_action(
+        ctx,
+        "warning_sd_retry",
+        "SD card problem",
+        action="None",
```
Comment (matejcik) [inline core/src/apps/common/sdcard.py:105]: didn't notice it before but this should probably be `None`, not `"None"`

Comment (matejcik) [reply core/src/apps/common/sdcard.py:105]: also means this is not covered by UI tests, correct?
that's probably because the SD card persistence tests don't have UI recording...

File: core/src/apps/webauthn/add_resident_credential.py [hunk 14]
```diff
@@ -41,15 +41,15 @@ async def add_resident_credential(
...
-            "not belong to this",
-            "authenticator.",
+        await show_error_and_raise(
+            ctx,
+            "warning_credential",
+            header="Import credential",
+            button="Close",
+            content="The credential you are trying to import does\nnot belong to this authenticator.",
+            red=True,
+            exc=wire.ActionCancelled,
```
Comment (matejcik) [inline core/src/apps/webauthn/add_resident_credential.py:51]: this is the default, no?

File: core/src/trezor/ui/components/common/confirm.py [hunk 15]
```diff
@@ -1,16 +1,26 @@
...
 CONFIRMED = object()
 CANCELLED = object()
 INFO = object()
 
 
+def is_confirmed(x: Any) -> bool:
+    return x is CONFIRMED
+
+
+async def raise_if_cancelled(a: Awaitable, exc: Any = wire.ActionCancelled) -> None:
```
Comment (matejcik) [inline core/src/trezor/ui/components/common/confirm.py:18]: the type should probably be `BaseException | Type[BaseException]` (here and in the other places that pass this through

**Review by matejcik — CHANGES_REQUESTED** [review]

File: core/src/apps/common/request_pin.py (hunk 11 above)
Comment (mmilata) [reply core/src/apps/common/request_pin.py:137]: It is but mypy needs it:
```
src/apps/common/request_pin.py:140: error: Implicit return in function which does not return  [misc]
```
Which is weird because `show_error_and_raise` is declared as `NoReturn`.

File: core/src/trezor/ui/components/common/confirm.py (hunk 15 above)
Comment (mmilata) [reply core/src/trezor/ui/components/common/confirm.py:18]: `wire.ActionCancelled` is an `int` though?

File: core/src/apps/common/request_pin.py (hunk 11 above)
Comment (matejcik) [reply core/src/apps/common/request_pin.py:137]: that's what I was thinking too. weird indeed. maybe something to do with asyncness of the function?
in any case let's put in an `assert False` instead to make it clearer for the reader

File: core/src/trezor/ui/components/common/confirm.py (hunk 15 above)
Comment (matejcik) [reply core/src/trezor/ui/components/common/confirm.py:18]: i'm somewhat sure it's an `Error`

File: core/src/apps/common/request_pin.py (hunk 11 above)
Comment (mmilata) [reply core/src/apps/common/request_pin.py]: Fixed in fef81373696d5dc2929b40dba0343afe855c9a7f. `assert False` does nothing in production firmware, right?

File: core/src/apps/common/request_pin.py (hunk 12 above)
Comment (mmilata) [reply core/src/apps/common/request_pin.py]: Fixed in fef81373696d5dc2929b40dba0343afe855c9a7f.

File: core/src/apps/webauthn/add_resident_credential.py (hunk 14 above)
Comment (mmilata) [reply core/src/apps/webauthn/add_resident_credential.py]: Fixed in fef81373696d5dc2929b40dba0343afe855c9a7f.

File: core/src/apps/common/request_pin.py (hunk 11 above)
Comment (matejcik) [reply core/src/apps/common/request_pin.py]: correct

File: core/src/trezor/ui/components/common/confirm.py (hunk 15 above)
Comment (mmilata) [reply core/src/trezor/ui/components/common/confirm.py:18]: D'oh, was looking `message/FailureType.py`. Maybe I should try something more intelligent than `ctags`.:thinking:. Fixed in fef81373696d5dc2929b40dba0343afe855c9a7f.

File: core/src/apps/common/sdcard.py (hunk 13 above)
Comment (mmilata) [reply core/src/apps/common/sdcard.py]: Oof, fixed in 82b690518f6c60f4d81c22421c45ad3e06322ee1. Will look into writing a testcase.

**Review by matejcik — APPROVED** [review]
the `sdcard` thing is looking much neater now :heart: 

i believe this is it. but let's discuss before merging if we want this to go to the upcoming release

File: core/src/apps/common/sdcard.py (hunk 13 above)
Comment (mmilata) [reply core/src/apps/common/sdcard.py]: There doesn't seem to be a simple way to inject SD card errors that lead to these dialogs:/ Checked manually that both work and look OK.

**Comment by mmilata** [comment]
So I've tested this with
```
core/emu.py --disable-animation --temporary-profile --heap-size=460K &
TESTOPTS="--random-order-bucket=none" make test_emu
```
and it works but with `--heap-size=450K` it fails on `test_15_of_15`. These two bounds are the same for master so it probably doesn't make the fragmentation much worse.

**Comment by matejcik** [comment]
let's get this merged

@mmilata please squash, rebase and fix conflicts

**Comment by mmilata** [comment]
Rebased, squashed, tests pass. Re-ran the heap sanity test, minimum `--heap-size` value is 459K for this branch and >=460K for master. Will merge this in the evening if there are no objections.
